### PR TITLE
os/bluestore/BlueFS: compact log on alloc runway fail

### PR DESCRIPTION
**Pull request by Aran85**
the log_writer extents may hit the end of BDEV_WAL when there are still some unused blocks.
so try to compact the log may resolve this situation, otherwise the OSD will failure forever.

tracker: #20645

Signed-off-by: Zengran <13121369189@126.com>

### CODE REVIEW

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -1391,7 +1393,15 @@ int BlueFS::_flush_and_sync_log(std::unique_lock<std::mutex>& l,
     int r = _allocate(log_writer->file->fnode.prefer_bdev,
 		      cct->_conf->bluefs_max_log_runway,
 		      &log_writer->file->fnode.extents);
+    if ((r == -ENOSPC) && _should_compact_log() && _try_num) {
+      // log_writer->file->fnode.extents may hint the end of the dev...
+      _compact_log_sync();
```
Comment (liewegas) [inline src/os/bluestore/BlueFS.cc]: I don't think it's safe to call _compat_log_sync() here.  See the first few lines of that function:
<pre>
  // clear out log (be careful who calls us!!!)
  log_t.clear();
</pre>
which will blow away at least something (given we just asserted that log_t is not empty).

That aside, I'm not sure this makes a lot of sense or that I understand the problem anyway.  Compacting will make the log more efficient, but this allocation will only fail if *all* devices are out of space... not just BDEV_WAL.  Which means you must be running dangerously close to full...?

Comment (Aran85) [inline src/os/bluestore/BlueFS.cc]: ohh yes...could we do the `::encode(log_t, bl);`  before call `_compact_log_sync()`  and lock the log_t? 
as indicated by tracker: #20645, EVEN the BDEV_WAL was not exhausted... the wal log file hold the first 400M of first 1G Zone, and the log_write'file(ino=1) hold the last 300MB of last 1G zone, not sure this is normal? when the log_write want extend its runway, but it hint the end of the BDEV_WAL(5GB), so allocator cant fail to other device? expecting your correct, Sir

**Comment by liewegas** [comment]
I have a feeling the reason this was deemed necessary was because we weren't compacting when we should.. see https://github.com/ceph/ceph/pull/17354.  If so, I think we can close this and not worry about this situation.  Making a retry will be fragile, hard to maintain, and would handle a case that should never happen.

**Comment by Aran85** [comment]
thanks, aggree with you... but i think in some situations `shutdown` maybe better than `assert`...

**Pull request closed without merging**